### PR TITLE
HTTP service layer for extractors

### PR DESCRIPTION
**Pull request by mfrisbey**
Not quite finished yet, but here is what I'm thinking - please feel free to look, play with it, and let me know what you think. I still need to update everything to react to your latest changes.

Here is what has changed:

* `HttpService`: new class that provides the HTTP interface for extractors. It takes the extractor it will use in its constructor. I added information about how to use it to the readme.
* `DefaultLogger`: potential default logger that everything can use when a logger isn't provided by the context. It uses a concept of used in the past, where it only logs messages to the console if an environment variable is set. Can be useful when running unit tests locally and you don't want to see all the log messages mixed in with the other output. LMK what you think.
* `ExtractProcess`: does the work of queueing the next page if the extractor provides more results. My thinking here is that we could reuse a class like this for a CLI, and it provides a single place where the queue logic exists; that way we can change one place if we decide to use SQS or something other than just calling the Lambda again. I couldn't tell if you already had something like this in your refactor. If we do, totally fine to remove this and use what you have.
* `LoggingSupport`: proposal for a base class that anything that needs to log messages can use. Centralizes the logging logic so that we don't need to duplicate it everywhere.
* `ExtractError`: this is the custom error class I wrote before I realized we had one in the rest stuff. I'll probably delete these and use the rest error instead.

### CODE REVIEW

File: src/logging-support.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * @type {Logger}
+   */
+  #log;
+
+  /**
+   * Constructs a new instance that will use the given configuration.
+   * @param {LoggerConfig} config Configuration for the logger.
+   */
+  constructor(config = {}) {
```
Comment (klcodanr) [inline src/logging-support.js:47]: Already available in the contexthelper in commons

File: src/http-service.js [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  /**
+   * Retrieves the ingestor URL associated with the current request.
+   * @param {import('@adobe/helix-universal').UniversalContext} context Context under
+   *  which the current execution is running.
+   * @returns {string} A URL.
+   */
+  static getIngestorUrl(context) {
+    const url = context.env.INGESTOR_URL;
+    if (!url) {
+      throw ExtractError.badRequest('Extractor services require an ingestor URL');
```
Comment (klcodanr) [inline src/http-service.js:121]: Is this really a bad request? This is an issue on our side.

File: src/http-service.js [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+   * Does the work of generating a response to a request, including handling any
+   * errors.
+   * @param {*} request An HTTP fetch request that was sent to the handler and
+   *  should be processed.
+   * @param {import('@adobe/helix-universal').UniversalContext} context Context under
+   *  which the current execution is running.
+   * @returns {HttpServiceResponse} The response that service will provide.
+   */
+  async generateResponse(request, context) {
+    const log = HttpService.getLogger(context);
```
Comment (klcodanr) [inline src/http-service.js:149]: Why not just pass the logger into this class?

File: src/http-service.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+      status = StatusCodes.INTERNAL_SERVER_ERROR,
+      title = StatusTexts.INTERNAL_SERVER_ERROR,
+      message: detail = DEFAULT_PROBLEM_DETAIL,
+    } = err;
+    const {
+      method,
+      url,
+    } = request;
+    HttpService.getLogger(context).error(`< ${method} ${status} ${url}: ${detail}`, err);
+    return {
```
Comment (klcodanr) [inline src/http-service.js:195]: Already covered in rest-error.js

File: src/http-service.js [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    try {
+      body = await request.json();
+    } catch (err) {
+      throw ExtractError.badRequest(`Invalid JSON received in request: ${err.message}`);
+    }
+
+    const { action } = body;
+
+    // extract is the only action that the service handles
+    if (action === ACTION_EXTRACT) {
```
Comment (klcodanr) [inline src/http-service.js:230]: Why even have an if statement if there is only one option?

**Review by klcodanr — CHANGES_REQUESTED** [review]
Mark,

I think there are a few things we need to work out here:

 - How do we handle authentication vs extraction? Should they be handled in one service or two?
 - Do we need a front layer for all of the extractors or should each extractor present an HTTP endpoint?
 - Do we handle the HTTP responses differently than the extraction jobs? The challenge is that the extraction cannot respond synchronously, so we need it to be invoked async where as the rest would all be standard HTTP responses.

File: src/default-logger.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+const LOG_LEVEL = process.env.EXTRACTOR_LOG_LEVEL;
+
+/**
+ * Class that will serve as the default logger for the module when none is provided.
+ * Will only write messages if a EXTRACTOR_LOG_LEVEL environment variable is
+ * set to the log level to write.
+ */
+export class DefaultLogger {
```
[the remark this answers is not in the document]
Comment (mfrisbey) [reply src/default-logger.js:20]: That's really the only reason I added the logger - for running locally. I've found it helpful to be able to turn the logging on/off when doing something like running unit tests. It keeps the results more clean and readable. I didn't intend this logger to go further than that, but I'd agree that we should look at winston if we do want to go further.

File: src/extract-process.js [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+   * @param {ExtractInfo} extractInfo Definition for the extraction to execute. This info may vary,
+   *  but must at least include the properties specified. It may include additional information
+   *  as required by individual extractors.
+   * @returns {Promise<*>} Resolves with the payload that was queued as the
+   *  next page. Will be falsy if there was not a next page.
+   */
+  async extract(extractInfo) {
+    let nextInfo = false;
+    const ingestorClient = this.getIngestorClient(extractInfo.process);
+    const nextPage = await ingestorClient.submitBatch(this.#getExtractor(), extractInfo, 1);
```
[the remark this answers is not in the document]
Comment (mfrisbey) [reply src/extract-process.js:154]: Yeah, still need to incorporate the latest changes into this branch. Will fix this as part of that.

File: src/http-service.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  /**
+   * Retrieves the ingestor API key associated with the current request.
+   * @param {import('@adobe/helix-universal').UniversalContext} context Context under
+   *  which the current execution is running.
+   * @returns {string} An API key.
+   */
+  static getIngestorApiKey(context) {
+    const apiKey = context.env.INGESTOR_API_KEY;
+    if (!apiKey) {
+      throw ExtractError.badRequest('Extractor services require an ingestor API key');
```
[the remark this answers is not in the document]
Comment (mfrisbey) [reply src/http-service.js:107]: Good point, I can make that change.

File: src/http-service.js (hunk 5 above)
Comment (mfrisbey) [reply src/http-service.js:230]: The stock extractor's lambda also allowed configuring it, using a different `action` value. I left this as a way of extending what the lambda can do.

**Pull request closed without merging**